### PR TITLE
Bump vispy

### PR DESCRIPTION
**Pull request by brisvag**
# Description
This PR bumps vispy to the latest, which allows us to get rid of our vendored code thanks to vispy/vispy#2227. It includes code from @alisterburt 's #3475 which fixed some issues with the new meshes.

EDIT: Moving some issue resolutions over from #3430 (which previously included this PR) to here.

# Related issues
This PR should solve the following issues:
- [wrong absolute size of points in data space](https://github.com/napari/napari/issues/2894#issuecomment-899888920). It won't solve the hardware limit to the marker size shown in #2894, though.
- #2637
- #3077
- #3301.
- #1310 is partially solved: no more errors, but NAN pixels are still opaque for some reason (despite being discarded). Will need to investigate further on the vispy side.

Not tested yet, but may be affected/related:
- ~#2213~ EDIT: this is still an issue
- ~#2138~ EDIT: this is still an issue

Also, clipping planes are now using world coordinates by default (this can be changed), previously data coordinates. See [this discussion](https://github.com/vispy/vispy/pull/2197#discussion_r706041780) for context.

# Notes
Something worth keeping in mind is that some things will look different for `Points`, due to the fact that after vispy 0.8.0 `edge_width` is calculated "correctly". This easy to fix, but might cause some confusion to people, especially if they are using very small points on a black canvas, which with the default black edge result in basically invisible points (i.e: too thck edge").

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [x] all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by brisvag** [comment]
I can't seem to force the CI to use the latest vispy main. For what it's worth, it works on my pc, so it should be ready to go :P

**Review by alisterburt — COMMENTED** [review]
so much red! Looks good, I'm looking at the problems noted in #3475 now and will PR here if I find a solution, I still think that should be fixed before merge - do you have any sense of timescale for the vispy release? Would you mind also giving a quick overview of any new features of note in vispy that this gains us? 🙂

**Comment by tlambert03** [comment]
yay oh yay!!

**Comment by brisvag** [comment]
> I'm looking at the problems noted in #3475 now and will PR here if I find a solution, I still think that should be fixed before merge

Yeah, I did not addressn that here yet, we should definitely do it before merge. We need to add a test for it as well!

> do you have any sense of timescale for the vispy release?

Not precisely; I @djhoese said he's gonna try to do a patch release with these changes, if possible (depending on other things that were merged in the meantime). Anyways, we still have some work to do here :)

> Would you mind also giving a quick overview of any new features of note in vispy that this gains us? slightly_smiling_face

Haven't checked out everything yet :P Here's some stuff from the top of my head:

- spherical points (molecules whoooo)
- clipping (though we vendored this already, there are some improvements, like using arbitrary coordinate systems)
- image/volume rendering should handle nans
- other stuff in meshes which probably goes way back and I don't understand :P

Some other changes were already implemented with vendored code, and are now included upstream (e.g: depth stuff by @kevinyamauchi ).

**Comment by djhoese** [comment]
VisPy 0.9.1 should be available.

**Comment by brisvag** [comment]
Thanks a lot @djhoese for the speedy release! Let's see how far the good old "works on my machine" goes :P

**Comment by brisvag** [comment]
So, the test currently failing is the one I already noticed when working on clipping planes and [trying to update vendored code](https://github.com/napari/napari/pull/3252#discussion_r698532395). At the time we just left it untouched and I opened an issue for it at #3301.

I think that there's something fishy with that test... Can someone more familiar with these tests chip in?

**Comment by alisterburt** [comment]
@brisvag I've done battle with that exact same test 😆 if you play with the attenuation threshold in the attenuated_mip rendering mode do things still behave as expected, i.e. things in the foreground become more visible? (easy to test with an EM map as an example) It may just be that the absolute values need changing but it could also be an issue with new vispy...

**Comment by brisvag** [comment]
As I pointed out in [the original comment](https://github.com/napari/napari/pull/3252#discussion_r698532395), to me it looks like this test is plain wrong, and we were testing for the wrong thing before:
```python
    ...
    viewer.layers[0].attenuation = 0.5
    screenshot = viewer.screenshot(canvas_only=True, flash=False)
    center = tuple(np.round(np.divide(screenshot.shape[:2], 2)).astype(int))
    # Check that rendering has not been attenuated
    assert screenshot[center + (0,)] > 80
```
Why is this check like this? Shouldn't this *be* attenuated, seeing as attenuation was set to 0.5?

Thing is, I don't know exactly what attenuation does to judge this.

**Comment by alisterburt** [comment]
The expected behaviour (visually) is that as you move further into the volume you downweight the contribution of the sample to the rendered pixel, implemented here prior to upstreaming/unvendoring https://github.com/napari/napari/blob/9af23fb0ddd8b9804edc84d534095a00b697c41c/napari/_vispy/vendored/volume.py#L508-L509

The visual effect worked nicely and if that's still the case then I'm fine with updating the fixed values in the tests to accommodate - I agree the test seems weird as written. 

If you're happy that the visual effect looks the same, maybe with slight differences in the image values for a given attenuation value, then we can update the test accordingly and move forward 🙂

**Comment by brisvag** [comment]
Yup, I can update the test! Just making sure I understand correctly first: what do you expect from an attenuation value of 0? And 1? That math is hard to parse :P

**Comment by brisvag** [comment]
After chatting a bit with @alisterburt and playing with it, my conclusion is that attenuation values are very unintuitive, and the slider is quite confusing. I changed the tests so they are consistent with how attenuation works *with this specific test data*, but nor this nor the older test would withstand data changes.

I propose we move forward like this (attenuation does the same thing as before, ~the curve is just different, apparently~ EDIT: actually, scratch that. The issue was that the attenuation was not updated correctly in some cases, which afaik was fixed by [this change, which we didn't vendor yet](https://github.com/vispy/vispy/blob/06bedb0e9aa410505dbe283d2c52dc9b168f8ded/vispy/visuals/volume.py#L1018-L1019)!), and I'll open a new, separete PR for streamlining what attenuation does (if that's even possible, given the math).

**Comment by brisvag** [comment]
So, the [*latest* test failure](https://github.com/napari/napari/pull/3494/checks?check_run_id=3962372045) comes from a bug I already noticed while working on this PR and for which I opened an issue upstream at vispy/vispy#2229. It was already fixed once with #1622, then upstreamed at vispy/vispy#1911. A slightly later PR (vispy/vispy#1920) introduced a regression by shuffling around some code.

@tlambert03, since you worked on the original fix, could you take a look?

PS: why is this bug only showing up wioth pip install? Shouldn't this fail everywhere?

**Comment by tlambert03** [comment]
I can, but kinda swamped right now so it will probably be a week or so

**Comment by jni** [comment]
@brisvag I don't really understand where all of the disparate discussions have landed. Could you please summarise? Some of the issues above that I don't know the resolution of:

- an attenuation test was failing but maybe the test was always wrong? I agree reading above that it looks weird. I think @sofroniewn implemented attenuated MIP originally so he might remember the rationale in the test, but my gut is that it was something quickly implemented so that we'd know if things changed, rather than something that needs to be preserved exactly.
- the attenuation curve may or may not have changed in recent vispy?
- the attenuation curve was not updating correctly — is this fixed? Is this fixed upstream? Is there an issue for this?
- the vispy issue on colortransforms — can that be fixed by vendoring again or should this PR just wait for that to be fixed on the vispy end?

**Comment by jni** [comment]
I've added this PR and #3301 to the agenda for the US/Pacific meeting tomorrow, in case anyone wants to advocate for particular outcomes/actions!

**Comment by brisvag** [comment]
# Attenuation

>     * an attenuation test was failing but maybe the test was always wrong? I agree reading above that it looks weird. I think @sofroniewn implemented attenuated MIP originally so he might remember the rationale in the test, but my gut is that it was something quickly implemented so that we'd know if things changed, rather than something that needs to be preserved exactly.

I checked again, and I was wrong. The reason why the test now fails is actually because of some kind of event slowness, similar to what we had with #3337.
In `ipython`, the way you split the calls matters. If you do the next, it will fail as in the tests:
```python
import napari
import numpy as np

viewer = napari.Viewer(ndisplay=3)
data = np.zeros((100, 10, 10))
data[-1] = 1
viewer.add_image(data, contrast_limits=[0, 1])
viewer.layers[0].rendering = 'attenuated_mip'
viewer.layers[0].attenuation = 0.5
screenshot = viewer.screenshot(canvas_only=True, flash=False)
center = tuple(np.round(np.divide(screenshot.shape[:2], 2)).astype(int))
assert screenshot[center + (0,)] > 80
```

However, if you split the calls like this, it won't fail:

```python
import napari
import numpy as np

viewer = napari.Viewer(ndisplay=3)
data = np.zeros((100, 10, 10))
data[-1] = 1
viewer.add_image(data, contrast_limits=[0, 1])
viewer.layers[0].rendering = 'attenuated_mip'
```
enter, then:
```python
viewer.layers[0].attenuation = 0.5
screenshot = viewer.screenshot(canvas_only=True, flash=False)
center = tuple(np.round(np.divide(screenshot.shape[:2], 2)).astype(int))
assert screenshot[center + (0,)] > 80
```

It's worth notign that vispy throws a warning that `u_attenuation` is unset.

The attenuation curve is actually unchanged, so it's just a matter of whether it was set or not. Apparently, the reason why my modified test passes is because of an addtional screenshot *before* the problematic lines (strangely, not *in between*). This works:
```python
import napari
import numpy as np

viewer = napari.Viewer(ndisplay=3)
data = np.zeros((100, 10, 10))
data[-1] = 1
viewer.add_image(data, contrast_limits=[0, 1])
### THE NEXT LINE IS THE ONLY DIFFERENT FROM BEFORE
screenshot = viewer.screenshot(canvas_only=True, flash=False)
viewer.layers[0].rendering = 'attenuated_mip'
viewer.layers[0].attenuation = 0.5
screenshot = viewer.screenshot(canvas_only=True, flash=False)
center = tuple(np.round(np.divide(screenshot.shape[:2], 2)).astype(int))
assert screenshot[center + (0,)] > 80
```

# Color transforms

>     * the vispy issue on colortransforms — can that be fixed by vendoring again or should this PR just wait for that to be fixed on the vispy end?

In theory, now it can be fixed by patching a subclass, rather than having to vendor. However, it's probably best to fix upstream.

**Comment by brisvag** [comment]
With 3321505 I worked around the issue @alisterburt found in [this comment](https://github.com/napari/napari/pull/3475#issuecomment-943159546). The reason why it happened with "big meshes" is actually that with all the small ones we tried before, vertices never ended up in 2D slices (since meshes don't have the n-dimensional thing that points have). So this was actually a general problem for meshes in 2D.

**Comment by alisterburt** [comment]
awesome work @brisvag, thanks for spelunking the attenuation test and dealing with filter attachment/detachment!

Looks like this is now only missing the vispy color transforms regression before we can merge? 🎉

**Comment by brisvag** [comment]
> Looks like this is now only missing the vispy color transforms regression before we can merge?

I'd say we should either find a solution for this weird race-condition-thingy with attenuation, or patch `Volume` to revert that upstream change so that `u_attenuation` is always define. Ohter than that, color transforms and then we're good!

**Comment by sofroniewn** [comment]
> I'd say we should either find a solution for this weird race-condition-thingy with attenuation, or patch Volume to revert that upstream change so that u_attenuation is always define. Ohter than that, color transforms and then we're good!

Can we try a PR to vispy to fix `u_attenuation` there. I think we need it to be the way we have it here. I remember this as a bug that i fixed for us. Is there a vispy PR with that change? I could then take a look.

they have this "if" statement https://github.com/vispy/vispy/blob/06bedb0e9aa410505dbe283d2c52dc9b168f8ded/vispy/visuals/volume.py#L1018-L1019

that we don't have 
https://github.com/napari/napari/blob/9af23fb0ddd8b9804edc84d534095a00b697c41c/napari/_vispy/vendored/volume.py#L1165

**Comment by brisvag** [comment]
> they have this "if" statement that we don't have

Yes, we could reintroduce it in vispy; however, why would this behave differently from `u_threshold` just above? It seems to me that the implementation is basically identical.

**Comment by sofroniewn** [comment]
> Yes, we could reintroduce it in vispy; however, why would this behave differently from u_threshold just above? It seems to me that the implementation is basically identical.

I wonder if that might give us problems too and we've just never tested it. It was definitely tricky to get this right i recall. I think we had bugs and this was where we landed as a fix. it might need some more rigorous testing.

Do you know if we drop the if statement does it work @brisvag? That might be first thing to test

**Comment by brisvag** [comment]
Yes, I am sure that the if statement is the culprit. This is what I wrote down in #3301, which I found when changing basically only that line. 

I did some digging: the if statement in `u_attenuation` was added by @alisterburt, but from the PR it seems that it was just "because it's the same in `u_threshold`". This, in turn, was added 6 years ago in commit ed069fb8, but I could not track down a PR for it (not sure how to do something like that other than the github search function).

From what I gather, that if clause is useless. I made a pr over at vispy: vispy/vispy#2242.

File: napari/_vispy/experimental/tiled_image_visual.py
Comment (GenevieveBuckley) [inline napari/_vispy/experimental/tiled_image_visual.py]: I'm sorry, what does `XXX broken` mean in this context?

File: examples/spheres.py
Comment (GenevieveBuckley) [inline examples/spheres.py]: Is this a cosmetic change? Or does a value of 10 not work well with this specific version of vispy, and why?

File: napari/_tests/test_with_screenshot.py
Comment (GenevieveBuckley) [inline napari/_tests/test_with_screenshot.py]: Can we keep this as a 3D image, please?

File: examples/spheres.py
Comment (brisvag) [reply examples/spheres.py]: This change simply makes the mesh bigger, enough to trigger [the bug that I fixed here](https://github.com/napari/napari/pull/3494#issuecomment-949568163). I only changed to have a quick way to test the bug, but I can revert and introduce an actual test!

File: napari/_tests/test_with_screenshot.py
Comment (brisvag) [reply napari/_tests/test_with_screenshot.py]: This is still 3D. Anyways, the change was just temporary and slipped in, so I'll revert.

File: examples/spheres.py
Comment (GenevieveBuckley) [reply examples/spheres.py]: +1 to adding a test specifically for that bug

I don't have a strong opinion on whether to keep or revert the change on this line. Just want to make sure we all knew why it was that way.

File: napari/_vispy/experimental/tiled_image_visual.py
Comment (brisvag) [reply napari/_vispy/experimental/tiled_image_visual.py]: Uh oh, I actually forgot about this change. Thanks for catching this. I updated the code, but I'm not very familiar with tiled rendering. Can someone who is please take a look at this?

Comment (sofroniewn) [reply napari/_vispy/experimental/tiled_image_visual.py]: Does @perlman know enough about this part of the code yet to check it out? If not I can take a look later.

Comment (perlman) [reply napari/_vispy/experimental/tiled_image_visual.py]: Hmmm. Interesting...

I see that the new `_build_color_transform` method looks at `self._data.ndim`, but I don't know how that variable is used given the actual data being loaded is through many small textures...

One quick work-around would be to move the existing `_build_color_transform()` into `TiledImageVisual`.

Regardless, I'll need to try out this branch with some sample code.

Comment (djhoese) [reply napari/_vispy/experimental/tiled_image_visual.py]: Just in case it inspires a napari solution: in my own application where I have "image tiles" stored in a single texture, I end up storing `self._data` in the ImageVisual as a custom class that only implements a `.ndim` and `.shape` attribute with the information I want instead of having to store actual numpy arrays that may or may not exist during visual creation time.

**Comment by GenevieveBuckley** [comment]
@brisvag there are some file conflicts here, could you merge main?

File: napari/_vispy/experimental/tiled_image_visual.py
Comment (sofroniewn) [reply napari/_vispy/experimental/tiled_image_visual.py]: Thanks @djhoese. Probably for a discussion in another place, but I would love it long term if we could contribute our "image tiles" back upstream to vispy one day. It's still in an "experimental" position in napari, but one day maybe!

Comment (brisvag) [reply napari/_vispy/experimental/tiled_image_visual.py]: @perlman If I understand your comment correctly, that's exactly what I did. I overloaded the `_build_color_transform` in `TiledImageVisual` so that it looks at `self.tile_shape` instead of `self._data`, just like it was doing before in `prepare_draw`.

**Comment by brisvag** [comment]
A heads up: before merging, we should wait for the next bugfix release of vispy. Due to the changes I made for this PR to happen, a couple of issues came up! They're pretty much solved now though :)

**Comment by GenevieveBuckley** [comment]
> A heads up: before merging, we should wait for the next bugfix release of vispy. Due to the changes I made for this PR to happen, a couple of issues came up! They're pretty much solved now though :)

Is there a rough time frame available for that?

**Comment by djhoese** [comment]
There is one more PR authored by @brisvag that needs to be finalized and then I can make a release.

**Comment by brisvag** [comment]
Also, before merging we still have to solve the [`clim`/visibility issue](https://github.com/napari/napari/pull/3494#issuecomment-948515016); hopefully @tlambert03 has some time for it soon :)

**Comment by brisvag** [comment]
@sofroniewn: Just tested what you mentioned [in this comment](https://github.com/vispy/vispy/pull/2239#issuecomment-953078137). Everything seems fine in napari as well!

https://user-images.githubusercontent.com/23482191/139109779-ab2e18eb-a95d-4ae6-b662-023d5e24c0e0.mp4

For context, this is how the same volume looks like in napari *BEFORE* that vispy fix:

https://user-images.githubusercontent.com/23482191/139110313-94cd61d7-1c94-4ad6-8104-ca04ec59ddf0.mp4

As you can see, it's a 3D only issue, and it's not a problem with the *position* ov the volume in space, but with the texture lookup.

**Comment by djhoese** [comment]
VisPy 0.9.3 is being built now: https://github.com/vispy/vispy/actions/runs/1391204348

**Comment by GenevieveBuckley** [comment]
Looks like it's out now: https://github.com/vispy/vispy/releases/tag/v0.9.3

We'll have to re-run the CI with vispy 0.9.3 and see the tests pass before merging. I'm very happy about this PR!

**Comment by jni** [comment]
@GenevieveBuckley I've pushed an empty commit to re-run CI. I'll give this a proper review shortly. 😃

**Comment by sofroniewn** [comment]
> @sofroniewn: Just tested what you mentioned in this comment. Everything seems fine in napari as well!

@brisvag now that vispy has been updated do we need to update napari? I suspect so otherwise we might be "fixing" something that's not an issue anymore, right?

**Comment by brisvag** [comment]
No, the test I showed above was already with the updated vispy (I just installed from vispy/main in the local environment).

**Comment by jni** [comment]
@brisvag this PR is failing reliably in ubuntu 3.8 pip install... I can't say that I understand why...

https://github.com/napari/napari/runs/4034031354?check_suite_focus=true#step:5:228

**Comment by sofroniewn** [comment]
> @brisvag this PR is failing reliably in ubuntu 3.8 pip install... I can't say that I understand why...

Do we know if this fail is specific to this PR? We've got a couple new fails popping up. How close to merge is this now? Anything left?

**Comment by brisvag** [comment]
@jni @sofroniewn that fail is indeed caused by this PR. Specifically, it's [this issue here](https://github.com/napari/napari/pull/3494#issuecomment-948515016), which is currently being worked on by @tlambert03 in vispy/vispy#2245.

If that gets merged and a vispy patch gets released, we can then bump the dependency and we should be good to go. It's a bit of a tricky one, so it might take some time to narrow down, and it might turn out that it's cause by somethign we do wrong in napari.

**Comment by brisvag** [comment]
So, I realised that changing `visible` on a layer triggers `set_data` because it triggers `refresh`.
```
    @visible.setter
    def visible(self, visibility):
        self._visible = visibility
        self.refresh()  # <---
        self.events.visible()
        if self.visible:
            self.editable = self._set_editable()
        else:
            self.editable = False
```

I think this is the reason why we have that clim issue in the first place. I tried removing it and indeed it fixed that test, ~though it broke several others~ EDIT: actually, are these failures even related?. Can someone explain why exactly a visibility change should trigger this `refresh`? EDIT2: Ok, I realised why. Without it, changing dims slice while a layer is invisible and then making it visible will leave the layer at the previous slice.

**Comment by sofroniewn** [comment]
> If that gets merged and a vispy patch gets released, we can then bump the dependency and we should be good to go. It's a bit of a tricky one, so it might take some time to narrow down, and it might turn out that it's cause by somethign we do wrong in napari.

Ok thanks for update!!

**Comment by Carreau** [comment]
Merged main branch in , that should fix some failures.

**Comment by tlambert03** [comment]
If the last thing we're waiting on here is the cpu texture bug and 2445, I think we should just move to float textures  in this PR too.  That is essentially a vispy bug, not a napari bug, and it doesn't happen with float textures. Once vispy decides how it wants to solve the problem, it will be fixed here too, but we don't need to test for it

**Comment by brisvag** [comment]
I fixed some conflicts and tried to switch over to GPU textures. Some errors arised, I'll take a look later.

**Comment by brisvag** [comment]
I'll be reverting the GPU testure change; it's not trivial and should be in a separate PR. Vispy should get updated soon with the latest fix, then we'll be good to go!

**Comment by djhoese** [comment]
Vispy 0.9.4 is on PyPI now. Waiting on conda-forge.

**Review by sofroniewn — APPROVED** [review]
LGTM! @brisvag the mac test is failing on this PR. I'm not sure if that is now failing again everywhere but @Carreau has been chasing down a lot of these fails lately and I don't want to introduce another one without realizing it. Can you try merging in main one more time if it isn't already done. Then I'd like to merge this as there is a bunch of other vispy related PRs that we have open that i think depend on it! Awesome work here, and thanks also to the vispy team for getting 0.9.4 out too!!

**Comment by sofroniewn** [comment]
Merging this now, we'll follow up with fixing mac test - see discussion in #3693

**Pull request merged**
